File: minecommons-paper/src/main/java/com/github/aecsocket/minecommons/paper/serializers/BlockDataSerializer.java
package com.github.aecsocket.minecommons.paper.serializers;

import org.bukkit.Bukkit;
import org.bukkit.block.data.BlockData;
import org.checkerframework.checker.nullness.qual.Nullable;
import org.spongepowered.configurate.ConfigurationNode;
import org.spongepowered.configurate.serialize.SerializationException;
import org.spongepowered.configurate.serialize.TypeSerializer;

import static com.github.aecsocket.minecommons.core.serializers.Serializers.require;

import java.lang.reflect.Type;

/**
 * Type serializer for a {@link BlockData}.
 * <p>
 * Uses {@link BlockData#getAsString(boolean)} and {@link Bukkit#createBlockData(String)}.
 */
public class BlockDataSerializer implements TypeSerializer<BlockData> {
    /** A singleton instance of this serializer. */
    public static final BlockDataSerializer INSTANCE = new BlockDataSerializer();

    @Override
    public void serialize(Type type, @Nullable BlockData obj, ConfigurationNode node) throws SerializationException {
        if (obj == null) node.set(null);
        else {
            node.set(obj.getAsString(true));
        }
    }

    @Override
    public BlockData deserialize(Type type, ConfigurationNode node) throws SerializationException {
        return Bukkit.createBlockData(require(node, String.class));
    }
}

File: paper/src/main/kotlin/com/github/aecsocket/alexandria/paper/serializer/BlockDataSerializer.kt
package com.github.aecsocket.alexandria.paper.serializer

import com.github.aecsocket.alexandria.core.extension.force
import org.bukkit.Bukkit
import org.bukkit.block.data.BlockData
import org.spongepowered.configurate.ConfigurationNode
import org.spongepowered.configurate.serialize.SerializationException
import org.spongepowered.configurate.serialize.TypeSerializer
import java.lang.reflect.Type

object BlockDataSerializer : TypeSerializer<BlockData> {
    override fun serialize(type: Type, obj: BlockData?, node: ConfigurationNode) {
        if (obj == null) node.set(null)
        else {
            node.set(obj.getAsString(true))
        }
    }

    override fun deserialize(type: Type, node: ConfigurationNode) = try {
        Bukkit.createBlockData(node.force<String>())
    } catch (ex: IllegalArgumentException) {
        throw SerializationException(node, type, "Invalid block data", ex)
    }
}
